File: ios/SwingCore/Tests/make_golden.py
"""Regenerate SwingCoreTests/Resources/golden_real_swing_01.json.

Runs the repository's real-swing fixture through the browser engine (node), which
swingml/tests hold to the Python pipeline, and writes every intermediate the Swift
tests compare against. Run from ios/SwingCore after exporting a new model.
"""

import json
import subprocess
import tempfile
from pathlib import Path

import numpy as np

HERE = Path(__file__).resolve().parent
REPO = HERE.parents[2]
fixture = np.load(REPO / "swingml" / "tests" / "fixtures" / "real_swing_01.npz")
job = {
    "xy": np.round(fixture["xy"].astype(float), 6).tolist(),
    "visibility": np.round(fixture["visibility"].astype(float), 6).tolist(),
    "world": np.zeros((fixture["xy"].shape[0], 33, 3)).tolist(),
    "detected": [bool(v) for v in fixture["detected"]],
    "times": [float(t) for t in fixture["timestamps_s"]],
    "width": int(fixture["frame_width"]),
    "height": int(fixture["frame_height"]),
}
payload = HERE.parent / "Sources" / "SwingCore" / "Resources" / "model.json"
with tempfile.NamedTemporaryFile("w", suffix=".json", delete=False) as handle:
    json.dump(job, handle)
done = subprocess.run(
    ["node", str(HERE / "make_golden.mjs"), handle.name, str(payload)],
    capture_output=True, text=True, check=True,
)
target = HERE / "SwingCoreTests" / "Resources" / "golden_real_swing_01.json"
target.write_text(done.stdout, encoding="utf-8")
print(f"wrote {target} ({target.stat().st_size / 1e6:.2f} MB)")

File: ios/SwingCore/Tests/make_golden.mjs
// Reference answers for the Swift tests, from the browser engine - which is held
// to the Python pipeline to five decimal places by swingml/tests. Regenerate with:
//   python ../../swingml/tests/fixtures/... (see make_golden.py next to this file)
import { readFileSync } from "node:fs";
import { PoseSequence, resamplePose, extractFeatures } from "../../../swingml/webapp/engine.js";
import { SwingEventModel, decodeEvents } from "../../../swingml/webapp/model.js";
import { computeMetrics } from "../../../swingml/webapp/metrics.js";

const job = JSON.parse(readFileSync(process.argv[2], "utf8"));
const payload = JSON.parse(readFileSync(process.argv[3], "utf8"));
const config = payload.features;
const sequence = new PoseSequence(job.xy, job.visibility, job.world, job.detected, job.times,
                                  job.width, job.height);
const { sequence: resampled } = resamplePose(sequence, config.canonical_rate_hz);
const net = new SwingEventModel(payload);
const out = { input: job, resampledTimes: resampled.times, cases: {} };
for (const hand of ["right", "left"]) {
  const { features, n, width } = extractFeatures(resampled, hand, config);
  const logits = net.forward(features, n, width);
  const decoded = decodeEvents(logits, n, payload.architecture.classes,
    payload.thresholds.min_mean_confidence, payload.thresholds.min_core_confidence || 0);
  const rows = [];
  for (let t = 0; t < n; t += 7) rows.push({ t, values: Array.from(features.slice(t * width, (t + 1) * width)) });
  const entry = { n, width, featureRows: rows, logits: Array.from(logits), ok: decoded.ok };
  if (decoded.ok) {
    Object.assign(entry, {
      frames: decoded.frames, subframe: decoded.subframe, confidence: decoded.confidence,
      meanConfidence: decoded.meanConfidence,
      metrics: computeMetrics(resampled, decoded, hand, config),
    });
  } else entry.reason = decoded.reason;
  out.cases[hand] = entry;
}
process.stdout.write(JSON.stringify(out));
